### ISSUE TITLE
Warning: Prop `id` did not match. Server: "popover-trigger-33" Client: "popover-trigger-6"

### ISSUE THREAD

**Issue by nghiepit** [chakra-ui/chakra-ui#3020]
Hi,

After upgrade to `@chakra-ui@1.1.4,` my application working on `Next.js` throw a warning error.

Maybe has some breaking changes, the internal `auto-id` hook instead of `@reach-ui/auto-id`.

![image](https://user-images.githubusercontent.com/4768095/104225262-bf772f80-5478-11eb-9ce8-8151501e15c4.png)

**Comment by santialbo** [comment]
I'm having the same problem with all components that use `auto-id`.
On my first request after starting the Next.js server things are OK.
Its' only after this first request that it looks like the id counter is not being reset and the generated ids from the server keep increasing indefinitely.

**Comment by hoangnguyen94** [comment]
This error occurs with menu-button-id as well.

**Comment by dlindenkreuz** [comment]
Same with `AccordionButton`:
```
Warning: Prop `id` did not match. Server: "accordion-button-5" Client: "accordion-button-2"
button
withEmotionCache/<@webpack-internal:///./.yarn/$$virtual/@emotion-react-virtual-227e34bb4a/0/cache/@emotion-react-npm-11.1.4-e22adbc3be-a75d76ba0b.zip/node_modules/@emotion/react/dist/emotion-element-4fbd89c5.browser.esm.js:44:73
AccordionButton<@webpack-internal:///./.yarn/$$virtual/@chakra-ui-accordion-virtual-87af051e12/0/cache/@chakra-ui-accordion-npm-1.0.4-67ef2c56d8-3491791d55.zip/node_modules/@chakra-ui/accordion/dist/esm/accordion.js:144:7
div
withEmotionCache/<@webpack-internal:///./.yarn/$$virtual/@emotion-react-virtual-227e34bb4a/0/cache/@emotion-react-npm-11.1.4-e22adbc3be-a75d76ba0b.zip/node_modules/@emotion/react/dist/emotion-element-4fbd89c5.browser.esm.js:44:73
AccordionItem<@webpack-internal:///./.yarn/$$virtual/@chakra-ui-accordion-virtual-87af051e12/0/cache/@chakra-ui-accordion-npm-1.0.4-67ef2c56d8-3491791d55.zip/node_modules/@chakra-ui/accordion/dist/esm/accordion.js:82:7
```

**Comment by tomdohnal** [comment]
This issue is only happening on a non-production build of Next.js
Nonetheless, according to what's written in the React UI auto-id package source code (https://github.com/reach/reach-ui/blob/develop/packages/auto-id/src/index.tsx), the "incrementing integer" approach which ChakraUI uses can also cause some problems with the concurrent mode in React.

**Comment by santialbo** [comment]
> This issue is only happening on a non-production build of Next.js
> Nonetheless, according to what's written in the React UI auto-id package source code (https://github.com/reach/reach-ui/blob/develop/packages/auto-id/src/index.tsx), the "incrementing integer" approach which ChakraUI uses can also cause some problems with the concurrent mode in React.
> 
> I'd suggest using the same approach as the auto-id package from Reach UI is using

The *warning* is only happening on development because it's a development warning but the issue is present on production builds (as long as the page is not statically generated of course).
Eventhough react rehydration missmatches are not a big issue, they shouldn't be happening since it impacts performance.

**Comment by mizhon** [comment]
I resolve this by setting the Next.js SSR to 'false' for the related components.
Codes like below:
`const TopNavBarWithNoSSR = dynamic(import('../components/TopNavBar'), {
  ssr: false
})
`

**Comment by santialbo** [comment]
Well this is not exactly resolving the issue. You are just avoiding it by
not rendering that component on the server.

**Comment by tomdohnal** [comment]
> > This issue is only happening on a non-production build of Next.js
> > Nonetheless, according to what's written in the React UI auto-id package source code (https://github.com/reach/reach-ui/blob/develop/packages/auto-id/src/index.tsx), the "incrementing integer" approach which ChakraUI uses can also cause some problems with the concurrent mode in React.
> > I'd suggest using the same approach as the auto-id package from Reach UI is using
> 
> The _warning_ is only happening on development because it's a development warning but the issue is present on production builds (as long as the page is not statically generated of course).
> Even though react rehydration missmatches are not a big issue, they shouldn't be happening since it impacts performance.

You're right, my bad...
The solution that Reach UI is using is to render all ids as `null` on the server and then use the increasing counter approach on the client (with `useLayoutEffect`). 

Essentially, it's something like this:
```javascript

let id = 0;
const genId = () => ++id;

function useId() {
  const [id, setId] = React.useState(null);

  useIsomorphicLayoutEffect(() => {
    if (id === null) {
      setId(genId());
    }
  }, []);

  return id != null ? String(id) : undefined;
}
```

Copy pasting the code from https://github.com/reach/reach-ui/blob/develop/packages/auto-id/src/index.tsx resolves the issue. 

I can make a PR if someone from the Chakra UI team approves

**Comment by segunadebayo** [comment]
Looking into this @nghiepit, thanks for investigating this @tomdohnal.

**Comment by f-t-alves** [comment]
Hey, guys. Tripped on this problem today as well when using FormLabel under `chakra-ui/react 1.6.4`.

**Comment by kousenlsn** [comment]
I'm also having this problem, using Next.js ``11.0.1``, ``chakra-ui/react 1.6.4``

![image](https://user-images.githubusercontent.com/14895129/124600952-ab1b5a80-de3d-11eb-9990-fdb3c45e087c.png)

**Comment by jigz** [comment]
Same here, `next 11.0.1` `@chakra-ui/react 1.6.4`
![Screenshot 2021-07-07 091134](https://user-images.githubusercontent.com/31827190/124685215-6659f700-df03-11eb-8756-bafacb66a157.png)

**Comment by FelipeRocha** [comment]
> Hey, guys. Tripped on this problem today as well when using FormLabel in NextJS under `chakra-ui/react 1.6.4`.

I just added the `key` parameter with a unique id
```xml
<FormLabel key={uniqueId} htmlFor={inputId}>"My label"</FormLabel>
```

**Comment by Davestring** [comment]
![Captura de Pantalla 2021-07-16 a la(s) 11 44 16](https://user-images.githubusercontent.com/15917800/125981339-1cad0d59-a94c-45bb-a3ae-947caf565cbd.png)

Same problem in Menu

**Comment by ricardo85x** [comment]
Same problem with menu too 

Version:

    "@chakra-ui/core": "^0.8.0",
    "@chakra-ui/react": "^1.6.5",

Error:

    Warning: Prop `id` did not match. Server: "menu-button-2" Client: "menu-button-4"
    at button

**Comment by tnmyk** [comment]
Me too :(

**Comment by MikeWillis** [comment]
This is happening to me as well on the Menu component, same as @ricardo85x 

This issue is marked as closed, can it be re-opened, or do we have to create a new issue?

**Comment by robsonkades** [comment]
I have same problem

**Comment by Emmanuel-Guerreiro** [comment]
I have the same problem with PopoverTrigger

**Comment by FiazAbrar** [comment]
same problem :( 
![Screenshot 2021-07-27 230354](https://user-images.githubusercontent.com/45792418/127197105-50358adc-7fc9-425c-aa33-bfd9787d50b9.png)

**Comment by yuyaamano23** [comment]
I have same problem, too.
Help,me!!

**Comment by Trimud** [comment]
Hi all,

I have faced the same issue with FormLabel component. As suggested here #4328 I have added manually the `id` and it doesn't throw an error.

Hope this helps anyone else.

**Comment by y471n** [comment]
I'm having the same issue with `<PinInputField />` within a `<PinInput>`
- Next: `11.1.0`
- Chakra: `1.6.6`

**Comment by marinvirdol** [comment]
I get the same error when using ```FormLabel``` 

![Screenshot 2021-09-02 at 22 30 36](https://user-images.githubusercontent.com/7512221/131911928-3dadb566-cce4-43a0-a0d0-dc5d213ae092.png)

- Next: ```11.1.0```
- Chakra: ```1.6.7```

**Comment by fedemotta** [comment]
Same error happening with Slider component:
```Warning: Prop `id` did not match. Server: "slider-track-2" Client: "slider-track-4"```
I manage to fix it adding a custom id.
```js
<Slider
value={form.minimum}
defaultValue={0.5}
step={0.05}
min={0.05}
max={1}
onChange={handleSlider}
onChangeStart={() => setShowSliderValue(true)}
onChangeEnd={() => setShowSliderValue(false)}
id="calculator-slider-1"
>
```

**Comment by Idrking** [comment]
Had the same issue with the `<MenuButton>` inside `<Menu>`
`Warning: Prop 'id' did not match. Server: "menu-button-2" Client: "menu-button-4"`

Solved by adding an explicit `id` to the parent `<Menu>`, not the `<MenuButton>` itself. 

- Next `11.1.0`
- Chakra/react `1.6.6`

**Comment by snaqu** [comment]
`<MenuButton as={Button} />` causes `Prop "id" did not match. Server: "menu-button-13" Client: "menu-button-26"`

- Chakra: `1.6.7`
- Next: `11.1.3-canary.15`

**Comment by maapteh** [comment]
Its everywhere the same. Next latest, Chakra latest:
```
Warning: Prop `id` did not match. Server: "accordion-button-1" Client: "accordion-button-2"
```
I dont like to give id's to DOM elements :(

**Comment by aquibbaig** [comment]
I think we should re-open this, this issue still persists

**Comment by whisher** [comment]
Same problem with
```
<MenuButton
          as={IconButton}
          aria-label="Options"
          icon={<MdShoppingCart />}
          variant="cart"
        />
```

**Comment by Jhappy77** [comment]
I can confirm this issue still exists with MenuButton in Menu that Snaqu and ldrking reported

I added keys and ids and the issue persisted. However, after deleting the menu that was causing the problem, rendering, and reinserting it (with keys/ids) there was no issue. Hopefully this helps anyone who is using ids as a manual stopgap until the issue is fixed

**Comment by parsher** [comment]
![image](https://user-images.githubusercontent.com/26663229/135210110-41639c7d-852a-421a-a348-77b3ea0ce1dd.png)

FormLabel has the same issue.

`<form onSubmit={handleSubmit(loginUser)}>
                <FormControl isInvalid={errors.name}>
                  <FormLabel htmlFor="name">name</FormLabel>
                  <Input
                    id="name"
                    placeholder="name"
                    {...register("name", {
                      required: "This is required",
                      minLength: {
                        value: 3,
                        message: "Minimum length should be 3",
                      },
                    })}
                  />
                  <FormErrorMessage>
                    {errors.name && errors.name.message}
                  </FormErrorMessage>
                </FormControl>
                <FormControl isInvalid={errors.password}>
                  <FormLabel htmlFor="name">password</FormLabel>
                  <Input
                    id="password"
                    placeholder="password"
                    {...register("password", {
                      required: "This is required",
                      minLength: {
                        value: 4,
                        message: "Minimum length should be 4",
                      },
                    })}
                  />
                  <FormErrorMessage>
                    {errors.password && errors.password.message}
                  </FormErrorMessage>
                </FormControl>
                <Button
                  mt={4}
                  colorScheme="teal"
                  isLoading={isSubmitting}
                  type="submit"
                >
                  Log in
                </Button>
              </form>`

**Comment by valdeniomarinho** [comment]
`Warning: Prop "id" did not match. Server: "popover-trigger-3" Client: "popover-trigger-6"`

- `"next": "11.1.2"`
- `"@chakra-ui/react": "^1.6.8"`

**Comment by EthanStandel** [comment]
Issue still persists, if I try to add an id to the Menu or the MenButton with the issue, it just appends my requested ID to the incorrectly generated ID.
![image](https://user-images.githubusercontent.com/16840301/136680875-1a1e8e41-365f-4dbc-920f-3ab1a1e26495.png)

**Comment by ffyuanda** [comment]
Same issue here...
Think we should reopen this?

`Warning: Prop 'id' did not match. Server: "field-1-label" Client: "field-2-label"`

Currently I am using -

- `"@chakra-ui/react": "^1.6.10"`
- `"next": "11.1.2"`

**Comment by woshixixi** [comment]
![image](https://user-images.githubusercontent.com/18056146/140614990-cb0a808d-fab8-4f42-ad05-d1f432cc951f.png)
me too , i use the IdProvider in _app.js too, but it still a warnning

**Comment by valentincostam** [comment]
> I solved this error by providing an ID for the parent component (Popover). Thus, Popover passes this ID on to the child components. In my scenario, I provided the ID "citites" to Popover and PopoverTrigger received the ID "popover-trigger-cities" automatically, not showing any more errors. The same situation happened for different components that have this kind of "subcomponent" structure, like FormControl for example. I hope this can help others.

Thanks, @ajvideira. That solved the problem for me.

**Comment by byron1st** [comment]
> Mine gone when I added `id` & `isLazy` to the `Menu`
> 
> ```js
>    <Menu id="navbar" isLazy>
>         <MenuButton
>           as={IconButton}
>           icon={<HamburgerIcon />}
>           aria-label="Options"
>         />
>     </Menu>
> ```

Thanks. It solved mine.

**Comment by ABBOS1994** [comment]
> 

Thank you

**Comment by leoncarey** [comment]
> Mine gone when I added `id` & `isLazy` to the `Menu`
> 
> ```js
>    <Menu id="navbar" isLazy>
>         <MenuButton
>           as={IconButton}
>           icon={<HamburgerIcon />}
>           aria-label="Options"
>         />
>     </Menu>
> ```

Solved here <3

**Comment by prolifel** [comment]
> Mine gone when I added `id` & `isLazy` to the `Menu`
> 
> ```js
>    <Menu id="navbar" isLazy>
>         <MenuButton
>           as={IconButton}
>           icon={<HamburgerIcon />}
>           aria-label="Options"
>         />
>     </Menu>
> ```

This solution not working on me. im currently using:
- "@chakra-ui/react": "^1.6.12"
- "next": "12.0.2"

**Comment by OptimBro** [comment]
Mine gone when adding random id to the component
![image](https://user-images.githubusercontent.com/52219366/146351900-6b693e55-414f-4f65-8e75-b5468d04bb27.png)

for the error in console...

![image](https://user-images.githubusercontent.com/52219366/146352007-488aeefa-6786-4b9a-ab57-51177bf59df6.png)

**Comment by devdev-dev** [comment]
Same problem here `isLazy` workaround was helpful, but hoping for React 18 to solve this issue fundamentally.

**Comment by lior-a** [comment]
adding isLazy to Menu and id solve the issue for me

**Comment by ngamsomset** [comment]
> > Mine gone when I added `id` & `isLazy` to the `Menu`
> > ```js
> >    <Menu id="navbar" isLazy>
> >         <MenuButton
> >           as={IconButton}
> >           icon={<HamburgerIcon />}
> >           aria-label="Options"
> >         />
> >     </Menu>
> > ```
> 
> Thanks. It solved mine.

This works

**Comment by christianhubbard** [comment]
Getting this issue on all form control components when using Next.js and chakra-ui 
"next": "12.0.7",
 "@chakra-ui/react": "^1.7.2",

**Comment by KathrynCrawford** [comment]
This issue is still happening for me even with and id and isLazy on my Menu components.

"next": "12.0.4",
"@chakra-ui/react": "1.7.2",

**Comment by samuliasmala** [comment]
FYI this issue is fixed in the [chakra-ui/react version 2](https://github.com/chakra-ui/chakra-ui/pull/6011) where the automatically generated id is using [the native `useId` hook of React 18](https://github.com/chakra-ui/chakra-ui/pull/6011/commits/98fe95e42ab845d9efe06639573e27db387d51b8) so no need to give custom id's anymore.
